core/src/main/scala/org/bitcoins/core/crypto/TxSigComponent.scala

```diff
@@ -342,12 +352,70 @@ sealed abstract class WitnessTxSigComponentRebuilt extends TxSigComponent {
     */
   def witnessScriptPubKey: WitnessScriptPubKey
 
-  override def sigVersion: SignatureVersion = SigVersionWitnessV0
+  override def sigVersion: SignatureVersion = witnessScriptPubKey match {
+    case _: WitnessScriptPubKeyV0 => SigVersionWitnessV0
+    case _: TaprootScriptPubKey   =>
+      //i believe we cannot have keypath spend here because we don't
+      //need to rebuild a spk with keypath spent
+      //https://github.com/bitcoin/bitcoin/blob/9e4fbebcc8e497016563e46de4c64fa094edab2d/src/script/interpreter.cpp#L399
+      SigVersionTapscript
+    case w: UnassignedWitnessScriptPubKey =>
+      sys.error(
+        s"Cannot determine sigVersion for an unassigned witness, got=$w")
+  }
 
   def witnessVersion: WitnessVersion = witnessScriptPubKey.witnessVersion
 
 }
 
+/** Tx sig component that contains the differences between BIP143 (segwit v0)
+  * transaction signature serialization and BIP341.
+  *
+  * The unique thing with BIP341 is the message commits to the scriptPubKeys
+  * of all outputs spent by the transaction, also
+  *
+  * If the SIGHASH_ANYONECANPAY flag is not set, the message commits to the amounts of all transaction inputs.[18]
+  *
+  * This means we need to bring ALL the outputs we are spending, even though this data structure
+  * is for checking the signature of a _single_ output.
+  *
+  * @see https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#common-signature-message
+  */
+case class TaprootTxSigComponent(
+    transaction: WitnessTransaction,
+    inputIndex: UInt32,
+    outputMap: Map[TransactionOutPoint, TransactionOutput],
+    flags: Seq[ScriptFlag])
+    extends WitnessTxSigComponent {
+  require(
+    scriptPubKey.isInstanceOf[TaprootScriptPubKey],
+    s"Can only spend taproot spks with TaprootTxSigComponent, got=$scriptPubKey")
```

Comment:
Would it instead make it better to make `WitnessTxSigComponent` to `WitnessTxSigComponent[T <: WitnessScriptPubKey]`?

Reply:
I tried doing this, this ended up being a huge diff on its own and ran into problems of having to use `WitnessTxSigComponent[_]` everywhere. I think this is the best path forward for now unless we go to Scala 3 :-)